File: Round 2 2009/D.cpp
// Problem: Watering Plants
// Language: C++
// Author: KirarinSnow
// Usage: g++ thisfile.cpp -o executable && ./executable <input.in >output.out


#include <math.h>
#include <stdio.h>
#include <iostream>

using namespace std;

int px[40];
int py[40];
int pr[40];
double cx[3500];
double cy[3500];

double out(int v, int w, double r)
{
    double xd = px[v]-cx[w];
    double yd = py[v]-cy[w];
    return sqrt(xd*xd + yd*yd) > r - pr[v] + 1.0e-6;
}

double solve()
{
    int n;
    cin >> n;
    for (int i = 0; i < n; i++)
    {
	cin >> px[i] >> py[i] >> pr[i];
    }

    double l = 0.0;
    double u = 708.0;
    while (u-l > 1.0e-6)
    {
	double r = (l+u)/2.0;

	int cc = 0;
	for (int c1 = 0; c1 < n; c1++)
	{
	    for (int c2 = c1+1; c2 < n; c2++)
	    {
		double x1 = px[c1];
		double x2 = px[c2];
		double y1 = py[c1];
		double y2 = py[c2];
		double r1 = pr[c1];
		double r2 = pr[c2];

		double x0 = sqrt((x1-x2)*(x1-x2) + (y1-y2)*(y1-y2));
		double xp = ((r-r1)*(r-r1) - (r-r2)*(r-r2))/2.0/x0 + x0/2.0;

		double yps = (r-r1)*(r-r1) - xp*xp;
		if (yps >= 0)
		{
		    double yp = sqrt(yps);
		    double cos = (x2-x1)/x0;
		    double sin = (y2-y1)/x0;
		   
		    cx[cc] = cos*xp - sin*yp + x1;
		    cy[cc] = cos*yp + sin*xp + y1;
		    cc++;

		    if (yps > 0)
		    {
			double yn = -yp;
			cx[cc] = cos*xp - sin*yn + x1;
			cy[cc] = cos*yn + sin*xp + y1;
			cc++;
		    } 
		}
	    }
	}

	for (int i = 0; i < n; i++)
	{
	    cx[cc] = px[i];
	    cy[cc] = py[i];
	    cc++;
	}

	for (int d = 0; d < cc; d++)
	{
	    for (int e = 0; e < cc; e++)
	    {
		bool f = true;
		for (int z = 0; z < n; z++)
		{
		    if (out(z, d, r) && out(z, e, r))
		    {
			f = false;
			break;
		    }
		}
		if (f) goto resume;
	    }
	}
	l = r;
	continue;

    resume:
	u = r;
    }

    return u;
}

int main()
{
    int cases;
    cin >> cases;

    for (int i = 1; i <= cases; i++)
    {
	printf("Case #%d: %0.6f\n", i, solve());
    }
    
    return 0;
}

File: Round 2 2009/D.py
#!/usr/bin/python
#
# Problem: Watering Plants
# Language: Python
# Author: KirarinSnow
# Usage: python thisfile.py <input.in >output.out
# Comments: Times out on large.


from math import *

def compute():
    def out(v, w):
        return sqrt((v[0]-w[0])**2 + (v[1]-w[1])**2) > w[2] - v[2] + 1e-6

    n = input()
    p = [map(int, raw_input().split()) for i in range(n)]

    l = min(map(lambda x: x[2], p))+0.0
    u = 708.0

    while u-l > 1.0e-6:
        r = (l+u)/2
        c = []
        for c1 in range(n):
            for c2 in range(c1+1, n):
                x1, y1, r1 = p[c1]
                x2, y2, r2 = p[c2]
                
                x0 = sqrt((x1-x2)**2 + (y1-y2)**2)
                xp = ((r-r1)**2 - (r-r2)**2)/2.0/x0 + x0/2.0
                try:
                    yp = sqrt((r-r1)**2 - xp**2)
                    yn = -yp
                    cos = (x2-x1)/x0
                    sin = (y2-y1)/x0

                    x = cos*xp - sin*yp + x1
                    y = cos*yp + sin*xp + y1
                    v = cos*xp - sin*yn + x1
                    w = cos*yn + sin*xp + y1
                    
                    c.append((x, y, r))
                    c.append((v, w, r))
                except:
                    pass
        
        c.extend(map(lambda z: (z[0], z[1], r), p))

        for d in c:
            for e in c:
                for z in p:
                    if out(z, d) and out(z, e):
                        break
                else:
                    break
            else:
                continue
            break
        else:
            l = r
            continue
        u = r
    
    return r

for i in range(input()):
    print "Case #%d: %0.6f" % (i+1, compute())
